### PR TITLE
Fix mutation of circuit when pulse scheduling (#3678)

### LINKED ISSUE #3678: Scheduling mutates the input quantum circuit
### Information

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?
The qiskit scheduler mutates the input circuit, [see here](https://github.com/Qiskit/qiskit-terra/blob/21beac232cf83e6ecc737f6b6755762c3a7de7c6/qiskit/scheduler/methods/basic.py#L96)

### What is the expected behavior?
The qiskit scheduler should not mutate the input circuit.

### Suggested solutions
Either remove this mutation, or copy the circuit before operating on.

### PR DESCRIPTION
**Pull request by michelle4654**
### Summary

Fixes #3678 by replacing the circuit with a copy of the circuit and then adding the barrier.

### Details and comments

@taalexander recommended creating a quantum circuit with a barrier and combining, but I found this seemed to have problems passing the test cases, so I instead made a copy of the circuit and added a barrier. Then that circuit was used for scheduling.

### CODE REVIEW

File: qiskit/scheduler/methods/basic.py [hunk 1]
```diff
@@ -93,7 +93,8 @@ def as_late_as_possible(circuit: QuantumCircuit,
     """
     sched = Schedule(name=circuit.name)
     # Align channel end times.
-    circuit.barrier()
+    circuit_barrier = circuit.copy()
```
Comment (taalexander) [inline qiskit/scheduler/methods/basic.py]: I think this is fine. There is no need to rename though. Could just do
```suggestion
    circuit = circuit.copy()
```

Comment (ajavadia) [reply qiskit/scheduler/methods/basic.py]: Actually I think what Thomas originally suggested can be done with this. Can you try this? I think it's faster than a deepcopy.
```suggestion
    circuit = circuit + QuantumCircuit(*circuit.qregs)
```

**Comment by ajavadia** [comment]
Can you add a test that scheduling does not mutate a circuit?

**Comment by michelle4654** [comment]
Thanks for all the help! I think I made the changes everyone asked for and added it in the latest commits. Sorry, contributing to open-source is a little new to me so lmk if I slipped up on any Github etiquette

File: test/python/scheduler/test_basic_scheduler.py [hunk 2]
```diff
@@ -614,3 +614,13 @@ def test_multiple_measure_in_3Q(self):
             (10, acquire(AcquireChannel(2), MemorySlot(1)))
         )
         self.assertEqual(sched.instructions, expected.instructions)
+
+    def test_alap_keeps_circuit_same(self):
```
Comment (ajavadia) [inline test/python/scheduler/test_basic_scheduler.py]:
```suggestion
    def test_scheduling_keeps_circuit_same(self):
```

File: test/python/scheduler/test_basic_scheduler.py [hunk 3]
```diff
@@ -614,3 +614,13 @@ def test_multiple_measure_in_3Q(self):
...
+        """Test that scheduling does not change the input circuit."""
+        q = QuantumRegister(2)
+        c = ClassicalRegister(2)
+        qc = QuantumCircuit(q, c)
+        qc.cx(q[0], q[1])
+        qc_sched = qc + QuantumCircuit(*qc.qregs)
```
Comment (ajavadia) [inline test/python/scheduler/test_basic_scheduler.py]: here it's better to explicitly copy to make sure we are really comparing to the original (unmodified) circuit
```suggestion
        qc_copy = qc.copy()
```

File: test/python/scheduler/test_basic_scheduler.py [hunk 4]
```diff
@@ -614,3 +614,13 @@ def test_multiple_measure_in_3Q(self):
...
+        schedule(qc_sched, self.backend, method="alap")
```
Comment (ajavadia) [inline test/python/scheduler/test_basic_scheduler.py]:
```suggestion
        schedule(qc, self.backend, method="alap")
```

File: test/python/scheduler/test_basic_scheduler.py [hunk 5]
```diff
@@ -614,3 +614,13 @@ def test_multiple_measure_in_3Q(self):
...
+        self.assertEqual(qc, qc_sched)
```
Comment (ajavadia) [inline test/python/scheduler/test_basic_scheduler.py]:
```suggestion
        self.assertEqual(qc, qc_copy)
```

**Comment by michelle4654** [comment]
Sorry, I got confused on what's happening with this -- it looks like someone fixed basic.py another way? Should (can) I close this?

**Comment by taalexander** [comment]
@michelle4654, it seems this was fixed indirectly by #3795. I apologize for the wasted effort here :( , your work was coming along nicely. If you would like some help finding another issue to work on, I could work with you to find something that you are interested in. We value all Qiskit contributors and I regret that this experience was confusing for you.

**Comment by michelle4654** [comment]
Hi @taalexander . Thank you so much!! Don't worry about it -- I probably bit off a little bit more than I could chew since I didn't have a deep understanding of the scheduler and how to best improve it. I just wanted to dive into something and get a sense of what contributions were like (which is good, now I know a little bit more about unit tests and the general contribution protocol). 

I would love some help in finding another issue to work on, though! I requested to push something yesterday on issue #3649 (as you can see by the mistake/reverted commit on this one), but that was just updating some warnings to users. I'm looking for something new and interesting to work on where I'm not stepping on anyone's toes -- any ideas?

**Comment by taalexander** [comment]
Hi @michelle4654, would you be interested in taking #3885 on? If you want to have a lower latency conversation feel free to reach out to me on our Slack.

**Pull request closed without merging**